src/virtual_machine_start.rs

```diff
@@ -0,0 +1,120 @@
+    let mut connection_reader = BufReader::new(&stream);
+
+    loop {
+        let mut command = String::new();
+        connection_reader
+            .read_line(&mut command)
+            .map_err(|error| error.to_string())?;
+        let command = command.trim();
+
+        match command {
+            "n" => {
+                virtual_machine.next_instruction();
+                let memory_address = virtual_machine.read_register(Register::ProgramCounter);
+                let instruction = virtual_machine.memory_read(memory_address);
+                let response = format!("instruction: {instruction:#018b}",);
+                writeln!(&stream, "{}", response).map_err(|error| error.to_string())?;
+            }
+            "r" => {
+                let response = virtual_machine.state_of_registers();
+                writeln!(&stream, "{}", response).map_err(|error| error.to_string())?;
+            }
+            _ => match command.parse::<u16>() {
+                Ok(amount_of_instructions) => {
+                    virtual_machine.next_instructions(amount_of_instructions);
+                    let response = format!("executed {} instructions", amount_of_instructions);
+                    writeln!(&stream, "{}", response).map_err(|error| error.to_string())?;
+                }
+                Err(_) => {
+                    writeln!(&stream, "Invalid Command").map_err(|error| error.to_string())?
+                }
+            },
+        }
+    }
+}
+
+fn print_instructions_for_interactive_console() {
+    println!("Starting interactive console");
+    println!("Instructions: ");
+    println!("<r> to print the state of the registers at the moment");
+    println!("<an instruction in binary> to instantly execute that instruction");
+}
+
+pub fn execute_vm_in_interactive_mode() -> Result<(), String> {
+    let program_counter_start = 0x3000;
+    let mut virtual_machine = LC3VirtualMachine::new(program_counter_start);
+    print_instructions_for_interactive_console();
+    for line in stdin().lock().lines() {
+        let line = line.map_err(|error| error.to_string())?;
+        match line.as_str() {
+            "r" => {
+                let registers = virtual_machine.state_of_registers();
+                let registers = registers.split(STREAM_DATA_SEPARATOR);
+                for register in registers {
+                    println!("{}", register);
+                }
+            }
+            _ => match u16::from_str_radix(&line, 2) {
+                Ok(instruction) => {
+                    virtual_machine.decode_instruction(instruction);
+                    let instruction = format!("{instruction:#018b}",);
+                    println!("instruction proccess: {}", instruction);
+                }
+                Err(_) => println!("Wrong instruction format"),
+            },
+        }
+    }
+    Ok(())
+}
+
+fn receive_file(arg: String) -> Result<BufReader<File>, String> {
```

Comment:
If you change this to:
```suggestion
fn receive_file(arg: impl AsRef<Path>) -> Result<BufReader<File>, String> {
```
You can avoid all the `.to_string` calls you add on the call site. Both `String`, `&str` and `Path` would just work.

Reply:
Just searched about the `AsRef` trait, didn't know about it and seems really interesting for generics